### PR TITLE
Add filter for excluding precache images

### PR DESCRIPTION
**Pull request by sudomakeinstall2**
Precaching pulls more images than it is needed. This story enables the user
to provide a list of patterns to be excluded from precaching. User can specify
"excludePrecachePatterns" field in the overrides configmap. Using the command
"grep -vG -f {patterns-file}" precaching will filter out the patterns specified.
Below is an example of override configmap with excludePrecachePatterns:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-group-upgrade-overrides
data:
  excludePrecachePatterns: |
    aws
    thanos
```

closes:
https://issues.redhat.com/browse/OCPBUGS-3092
https://issues.redhat.com/browse/CNF-6848

Signed-off-by: Saeid Askari <saskari@redhat.com>

### CODE REVIEW

File: controllers/precache.go [hunk 1]
```diff
@@ -304,6 +305,8 @@ func (r *ClusterGroupUpgradeReconciler) includeSoftwareSpecOverrides(
 	}
 	rv.OperatorsPackagesAndChannels = operatorsPackagesAndChannels
 
+	rv.ExcludePrecachePatterns = strings.Split(overrides["excludePrecachePattern"], "\n")
+
```
Comment (vitus133) [inline controllers/precache.go:309]: Same here - IMHO not needed

**Review by vitus133 — COMMENTED** [review]
Added a couple of considerations for using the overrides configmap. General notes:
1. It is better to provide extended commit message with a body explaining reasoning behind the change (can include references to JIRA)
2. Did we check that upgrade succeeds if some of the images have been omitted? Maybe our changes alone are not enough?

File: config/pre-cache/overrides.yaml [hunk 2]
```diff
@@ -3,6 +3,9 @@ kind: ConfigMap
 metadata:
   name: cluster-group-upgrade-overrides
 data:
+  excludePrecachePatterns: |
+    aws
+    thanos
```
[the remark this answers is not in the document]
Comment (sudomakeinstall2) [reply config/pre-cache/overrides.yaml:8]: You are right, this was a mistake on my part.

[the remark this answers is not in the document]
Comment (jc-rh) [reply config/pre-cache/overrides.yaml:8]: Hmm, I am not following. The precaching spec structure name is a bit misleading. It's actually part of the CGU status. How can it be set if it's not in the overrides cm?

[the remark this answers is not in the document]
Comment (sudomakeinstall2) [reply config/pre-cache/overrides.yaml:8]: I thought you meant it should be commented like other fields? Is that what you mean? or do you mean it should not be overrdieable at all?

File: config/pre-cache/overrides.yaml [hunk 3]
```diff
@@ -13,4 +13,6 @@ data:
   #   performance-addon-operator: 4.9
   #   ptp-operator: stable
   #   sriov-network-operator: stable
-
+  # excludePrecachePatterns: |
+  #   aws
```
Comment (jc-rh) [inline config/pre-cache/overrides.yaml:17]: @vitus133 @imiller0 Any other common patterns should be included in the example?

File: config/pre-cache/overrides.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (vitus133) [reply config/pre-cache/overrides.yaml]: Sorry @jc-rh you are right, I got confused.
Withdrawing all comments on this topic

File: pre-cache/test.sh [hunk 4]
```diff
@@ -21,16 +21,25 @@ echo "test_index1" > /tmp/operators.indexes
 echo "test_index2" >> /tmp/operators.indexes
 echo "  package1:  channel1 " > /tmp/operators.packagesAndChannels
 echo "  package2:channel2" >> /tmp/operators.packagesAndChannels
+echo "aws" > /tmp/excludePrecachePatterns
```
Comment (jc-rh) [inline pre-cache/test.sh:24]: Add trailing empty lines

**Review by jc-rh — APPROVED** [review]

**Pull request merged**